library/include/rocwmma/internal/b32_util.hpp

```diff
@@ -0,0 +1,199 @@
+    template <>
+    struct b32Util<float64_t>
+    {
+        using ExtractedT = float32_t;
+        using EmplacedT  = float64_t;
+
+        using FootSpace = union
+        {
+            ExtractedT extracted[2];
+            EmplacedT  emplaced;
+        };
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto extract(VecT<EmplacedT, VecSize> const& v)
+        {
+            VecT<ExtractedT, VecSize> result;
+            auto const                rIt = makeVectorIterator(v).begin();
+            auto                      wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(EmplacedT) / sizeof(ExtractedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.emplaced   = get<0>(*rIt);
+                get<0>(*wIt) = a.extracted[SelectIdx];
+            }
+            return result;
+        }
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto emplace(VecT<ExtractedT, VecSize> const& v)
+        {
+            VecT<EmplacedT, VecSize> result;
+            auto const               rIt = makeVectorIterator(v).begin();
+            auto                     wIt = makeVectorIterator(result).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(EmplacedT) / sizeof(ExtractedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.extracted[SelectIdx] = get<0>(*rIt);
+                get<0>(*wIt)           = a.emplaced;
+            }
+            return result;
+        }
+
+        template <uint32_t VecSize, uint32_t SelectIdx = 0u>
+        __device__ static auto emplace(VecT<EmplacedT, VecSize>&        dst,
+                                       VecT<ExtractedT, VecSize> const& v)
+        {
+            auto const rIt = makeVectorIterator(v).begin();
+            auto       wIt = makeVectorIterator(dst).begin();
+
+            static_assert(decltype(rIt)::range() == decltype(wIt)::range(),
+                          "Unexpected iterator range mismatch");
+
+            static_assert(SelectIdx < (sizeof(EmplacedT) / sizeof(ExtractedT)),
+                          "Invalid index selection");
+
+            for(uint32_t i = 0u; i < decltype(rIt)::range(); i++, rIt++, wIt++)
+            {
+                FootSpace a;
+                a.extracted[SelectIdx] = get<0>(*rIt);
+                get<0>(*wIt)           = a.emplaced;
+            }
```

Comment:
Here we should initialize a.emplaced with get<0>(dst) first as well to preserve previous data